File: tests/pkTestData.ts
// Represents an id column, for a PK for instance
import {ColumnMap, ISourceMeta, PkMap, PKType} from "../src/jsonSchemaInspector"
import {Value} from "../src/SchemaTranslator"
import {uselessValueExtractor} from "./helpers"

export const id: PkMap = {
  prop: "id",
  sqlIdentifier: "`id`",
  chType: "UInt32",
  valueExtractor: (data) => parseInt(data.id),
  nullable: false,
  pkType: PKType.CURRENT,
  lowCardinality: false,
}

export const exhaustivePks: PkMap[] = [id, {
  ...id,
  sqlIdentifier: "`name`",
  chType: "String",
  prop: "name",
  valueExtractor: (data) => String(data.name),
}]

// Represents an id column, for a PK for instance
const rootId: PkMap = {
  prop: "_root_id",
  sqlIdentifier: "`_root_id`",
  chType: "UInt32",
  valueExtractor: (data) => parseInt(data._root_id),
  nullable: false,
  pkType: PKType.ROOT,
  lowCardinality: false,
}
// Represents a name column, as a simple column
const name: ColumnMap = {
  nullable: true,
  prop: "name",
  sqlIdentifier: "`name`",
  chType: "String",
  valueExtractor: (data) => data.name,
  lowCardinality: false,
}
// Represents a name column, as a simple column
export const valid: ColumnMap = {
  nullable: false,
  prop: "valid",
  sqlIdentifier: "`valid`",
  chType: "UInt8",
  valueExtractor: (data) => data.valid,
  valueTranslator: (v: Value) => {
    if (v === "true" || v === true || v === 1) {
      return 1
    } else {
      return 0
    }
  },
  lowCardinality: false,
}
export const simpleMeta: ISourceMeta = {
  pkMappings: [],
  simpleColumnMappings: [id, name],
  children: [],
  sqlTableName: "`order`",
  prop: "order",
}
const levelColumn = (lvl: number): PkMap => ({
  prop: `_level_${lvl}_index`,
  sqlIdentifier: `\`_level_${lvl}_index\``,
  chType: "UInt32",
  valueExtractor: () => {
    throw "should never be called"
  },
  nullable: false,
  pkType: PKType.LEVEL,
  lowCardinality: false,
})
export const metaWithPKAndChildren: ISourceMeta = {
  prop: "order",
  pkMappings: [id],
  simpleColumnMappings: [name],
  children: [{
    simpleColumnMappings: [name],
    pkMappings: [
      rootId, levelColumn(0),
    ],
    sqlTableName: "`order__tags`",
    prop: "tags",
    children: [{
      prop: "values",
      sqlTableName: "`order__tags__values`",
      pkMappings: [rootId, levelColumn(0), levelColumn(1)],
      simpleColumnMappings: [name],
      children: [],
    }],
  }],
  sqlTableName: "`order`",
}
export const abort = (err: Error) => {
  throw err
}
export const metaWithNestedValueArray: ISourceMeta = {
  prop: "audits",
  sqlTableName: "`audits`",
  pkMappings: [],
  simpleColumnMappings: [],
  children: [
    {
      prop: "events",
      sqlTableName: "`audits__events`",
      pkMappings: [
        {
          prop: "_level_0_index",
          sqlIdentifier: "`_level_0_index`",
          chType: "Int32",
          valueExtractor: uselessValueExtractor,
          nullable: false,
          pkType: PKType.LEVEL,
          lowCardinality: false,
        },
      ],
      simpleColumnMappings: [],
      children: [
        {
          prop: "previous_value",
          sqlTableName: "`audits__events__previous_value`",
          pkMappings: [
            {
              prop: "_level_0_index",
              sqlIdentifier: "`_level_0_index`",
              chType: "Int32",
              valueExtractor: uselessValueExtractor,
              nullable: false,
              pkType: PKType.LEVEL,
              lowCardinality: false,
            },
            {
              prop: "_level_1_index",
              sqlIdentifier: "`_level_1_index`",
              chType: "Int32",
              valueExtractor: uselessValueExtractor,
              nullable: false,
              pkType: PKType.LEVEL,
              lowCardinality: false,
            },
          ],
          simpleColumnMappings: [
            {
              sqlIdentifier: "`value`",
              valueExtractor: (data) => data,
              chType: "String",
              nullable: false,
              lowCardinality: false,
            },
          ],
          children: [],
        },
      ],
    },
  ],
}

File: src/test/kotlin/com/biron/singerTargetClickhouse/PkTestData.kt
package com.biron.singerTargetClickhouse

internal val id = PkMap(
	prop = "id",
	sqlIdentifier = "`id`",
	chType = "UInt32",
	valueExtractor = { (it as Map<*, *>)["id"]?.toString()?.toIntOrNull() },
	schemaType = null,
	typeFormat = null,
	nullable = false,
	lowCardinality = false,
	nestedArray = false,
	pkType = PKType.CURRENT,
)

private val rootId = PkMap(
	prop = "_root_id",
	sqlIdentifier = "`_root_id`",
	chType = "UInt32",
	valueExtractor = { (it as Map<*, *>)["_root_id"]?.toString()?.toIntOrNull() },
	schemaType = null,
	typeFormat = null,
	nullable = false,
	lowCardinality = false,
	nestedArray = false,
	pkType = PKType.ROOT,
)

private val name = ColumnMap(
	prop = "name",
	sqlIdentifier = "`name`",
	chType = "String",
	valueExtractor = { (it as? Map<*, *>)?.get("name") },
	schemaType = null,
	typeFormat = null,
	nullable = true,
	lowCardinality = false,
	nestedArray = false,
)

internal val validColumn = ColumnMap(
	prop = "valid",
	sqlIdentifier = "`valid`",
	chType = "UInt8",
	valueExtractor = { (it as? Map<*, *>)?.get("valid") },
	schemaType = "boolean",
	typeFormat = null,
	nullable = false,
	lowCardinality = false,
	nestedArray = false,
)

internal val idAsColumn = ColumnMap(
	prop = "id",
	sqlIdentifier = "`id`",
	chType = "Int32",
	valueExtractor = { (it as? Map<*, *>)?.get("id") },
	schemaType = null,
	typeFormat = null,
	nullable = false,
	lowCardinality = false,
	nestedArray = false,
)

private fun levelColumn(level: Int) = PkMap(
	prop = "_level_${level}_index",
	sqlIdentifier = "`_level_${level}_index`",
	chType = "Int32",
	valueExtractor = { throw IllegalStateException("level extractor should never be called") },
	schemaType = null,
	typeFormat = null,
	nullable = false,
	lowCardinality = false,
	nestedArray = false,
	pkType = PKType.LEVEL,
)

internal val simpleMeta = SourceMeta(
	prop = "order",
	sqlTableName = "`order`",
	pkMappings = emptyList(),
	simpleColumnMappings = listOf(idAsColumn, name),
	children = emptyList(),
)

internal val metaWithPKAndChildren = SourceMeta(
	prop = "order",
	sqlTableName = "`order`",
	pkMappings = listOf(id),
	simpleColumnMappings = listOf(name),
	children = listOf(
		SourceMeta(
			prop = "tags",
			sqlTableName = "`order__tags`",
			pkMappings = listOf(rootId, levelColumn(0)),
			simpleColumnMappings = listOf(name),
			children = listOf(
				SourceMeta(
					prop = "values",
					sqlTableName = "`order__tags__values`",
					pkMappings = listOf(rootId, levelColumn(0), levelColumn(1)),
					simpleColumnMappings = listOf(name),
					children = emptyList(),
				),
			),
		),
	),
)

internal val abort: (Throwable) -> Unit = { err -> throw err }

/**
 * Test helper: round-trip an in-memory [data] through the stream's real [StreamReader]
 * (serialize to JSON, re-parse). Uses the same decoder that the production hot path uses,
 * so tests can't silently diverge from it.
 */
internal fun mapToRow(meta: SourceMeta, data: Any?, translateValues: Boolean = false): RecordRow {
	val reader = StreamReader.from(meta, translateValues)
	val mapper = com.fasterxml.jackson.module.kotlin.jsonMapper {
		addModule(com.fasterxml.jackson.module.kotlin.kotlinModule())
	}
	val json = mapper.writeValueAsString(data)
	mapper.factory.createParser(json).apply { codec = mapper }.use { p ->
		p.nextToken()
		return reader.read(p)
	}
}

private val valueAsSelf = ColumnMap(
	prop = null,
	sqlIdentifier = "`value`",
	chType = "String",
	valueExtractor = { it },
	schemaType = null,
	typeFormat = null,
	nullable = false,
	lowCardinality = false,
	nestedArray = false,
)

internal val metaWithNestedValueArray = SourceMeta(
	prop = "audits",
	sqlTableName = "`audits`",
	pkMappings = emptyList(),
	simpleColumnMappings = emptyList(),
	children = listOf(
		SourceMeta(
			prop = "events",
			sqlTableName = "`audits__events`",
			pkMappings = listOf(levelColumn(0)),
			simpleColumnMappings = emptyList(),
			children = listOf(
				SourceMeta(
					prop = "previous_value",
					sqlTableName = "`audits__events__previous_value`",
					pkMappings = listOf(levelColumn(0), levelColumn(1)),
					simpleColumnMappings = listOf(valueAsSelf),
					children = emptyList(),
				),
			),
		),
	),
)
